python/ray/autoscaler/azure/config.py

```diff
@@ -0,0 +1,262 @@
+    except StopIteration:
+        logger.info("Creating MSI User Assigned Identity: %s", MSI_NAME)
+        identity = msi_client.user_assigned_identities.create_or_update(
+            resource_group_name=resource_group,
+            resource_name=MSI_NAME,
+            location=location)
+
+    identity_id = identity.id
+    principal_id = identity.principal_id
+    config["provider"]["msi_identity_id"] = identity_id
+    config["provider"]["msi_identity_principal_id"] = principal_id
+
+    # assign Contributor role for MSI User Identity to resource group
+    role_id = auth_client.role_definitions.list(
+        scope=resource_group_id, filter="roleName eq 'Contributor'").next().id
+    role_params = {"role_definition_id": role_id, "principal_id": principal_id}
+
+    for _ in range(RETRIES):
+        try:
+            filter_expr = "principalId eq '{}'".format(principal_id)
+            assignments = auth_client.role_assignments.list_for_scope(
+                scope=resource_group_id, filter=filter_expr)
+
+            if any(a.role_definition_id == role_id for a in assignments):
+                break
+
+            auth_client.role_assignments.create(
+                scope=resource_group_id,
+                role_assignment_name=uuid.uuid4(),
+                parameters=role_params)
+            logger.info("Assigning Contributor Role to MSI User")
+        except CloudError as ce:
+            if ce.inner_exception.error == "PrincipalNotFound":
+                time.sleep(5)
+    else:
+        raise Exception(
+            "Failed to create contributor role assignment (timeout)")
+
+    return config
+
+
+def _configure_key_pair(config):
+    ssh_user = config["auth"]["ssh_user"]
+    private_key_path = config["auth"].get("ssh_private_key")
+    if private_key_path:
+        # skip key generation if it is manually specified
+        assert os.path.exists(private_key_path), (
+            "Could not find private ssh key: {}".format(private_key_path))
+
+        # make sure public key also exists
+        public_key_path = config["auth"]["ssh_public_key"]
+        assert os.path.exists(public_key_path), (
+            "Could not find public ssh key: {}".format(public_key_path))
+    else:
+        resource_group = config["provider"]["resource_group"]
+
+        # look for an existing key pair
+        key_name = "ray_azure_{}_{}".format(resource_group, ssh_user)
+        public_key_path = os.path.expanduser("~/.ssh/{}.pub".format(key_name))
+        private_key_path = os.path.expanduser("~/.ssh/{}.pem".format(key_name))
+
+    if os.path.exists(public_key_path) and os.path.exists(private_key_path):
+        logger.info("Found SSH key pair: %s", key_name)
+        with open(public_key_path, "r") as f:
+            public_key = f.read()
+    else:
+        public_key, private_key_path = _generate_ssh_keys(key_name)
+        logger.info("Creating SSH key pair: %s", key_name)
```

Comment:
Is there a reason why we need to generate the ssh keys automatically? 

We do it for boto/aws because each instance needs to be associated with a particular registered keypair. Because that doesn't seem to be the case, I think you should just raise an informative/actionable error message if the private/public key isn't specified

Reply:
Yeah, I'm fine with having instructions on how to generate the ssh keys and leaving it up to the user before hand. Just wasn't sure what the expectation was here.

Reply:
OK sounds good!